File: iosApp/iosApp/blockchain/structs/SwapStruct.swift
//
//  SwapStruct.swift
//  iosApp
//
//  Created by jimmyt on 8/14/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import BigInt
import web3swift

/**
 Represents an on-chain [Swap](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#swap) struct.
 */
struct SwapStruct {
    /**
     Corresponds to an on-chain Struct's `isCreated` property.
     */
    let isCreated: Bool
    /**
     Corresponds to an on-chain Struct's `requiresFill` property.
     */
    let requiresFill: Bool
    /**
     Corresponds to an on-chain Struct's `maker` property.
     */
    let maker: EthereumAddress
    /**
     Corresponds to an on-chain Struct's `makerInterfaceId` property.
     */
    let makerInterfaceID: Data
    /**
     Corresponds to an on-chain Struct's `taker` property.
     */
    let taker: EthereumAddress
    /**
     Corresponds to an on-chain Struct's `takerInterfaceId` property.
     */
    let takerInterfaceID: Data
    /**
     Corresponds to an on-chain Struct's `stablecoin` property.
     */
    let stablecoin: EthereumAddress
    /**
     Corresponds to an on-chain Struct's `amountLowerBound` property.
     */
    let amountLowerBound: BigUInt
    /**
     Corresponds to an on-chain Struct's `amountUpperBound` property.
     */
    let amountUpperBound: BigUInt
    /**
     Corresponds to an on-chain Struct's `securityDepositAmount` property.
     */
    let securityDepositAmount: BigUInt
    /**
     Corresponds to an on-chain Struct's `takenSwapAmount` property.
     */
    let takenSwapAmount: BigUInt
    /**
     Corresponds to an on-chain Struct's `serviceFeeAmount` property.
     */
    let serviceFeeAmount: BigUInt
    /**
     Corresponds to an on-chain Struct's `serviceFeeRate` property.
     */
    let serviceFeeRate: BigUInt
    /**
     Corresponds to an on-chain Struct's `direction` property.
     */
    let direction: BigUInt
    /**
     Corresponds to an on-chain Struct's `settlementMethod` property.
     */
    let settlementMethod: Data
    /**
     Corresponds to an on-chain Struct's `protocolVersion` property.
     */
    let protocolVersion: BigUInt
    /**
     Corresponds to an on-chain Struct's `isPaymentSent` property.
     */
    let isPaymentSent: Bool
    /**
     Corresponds to an on-chain Struct's `isPaymentReceived` property.
     */
    let isPaymentReceived: Bool
    /**
     Corresponds to an on-chain Struct's `hasBuyerClosed` property.
     */
    let hasBuyerClosed: Bool
    /**
     Corresponds to an on-chain Struct's `hasSellerClosed` property.
     */
    let hasSellerClosed: Bool
    /**
     Corresponds to an on-chain Struct's `disputeRaiser` property.
     */
    let disputeRaiser: BigUInt
    
    /**
     Creates a `SwapStruct` from the `[AnyObject]` obtained via a web3swift read transaction call to [getSwap](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#get-swap).
     
     - Parameters:
        - response: The `[AnyObject]` obtained by a read transaction call to getSwap.
        - chainID: The ID of the blockchain from which `response` was obtained.
     
     - Returns: A new `SwapStruct` with property values derived from `response` and `chainID`, or `nil` if `response`'s `isCreated` property is `false`.
     
     - Throws: `BlockchainServiceError.unexpectedNilError` if an expected parameter in `response` is missing.
     */
    static func createFromGetSwapResponse(response: [AnyObject], chainID: BigUInt) throws -> SwapStruct? {
        guard let isCreated: Bool = response[0] as? Bool else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "isCreated"))
        }
        if !isCreated {
            return nil
        }
        guard let requiresFill = response[1] as? Bool else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "requiresFill"))
        }
        guard let maker = response[2] as? EthereumAddress else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "maker"))
        }
        guard let makerInterfaceID = response[3] as? Data else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "makerInterfaceID"))
        }
        guard let taker = response[4] as? EthereumAddress else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "taker"))
        }
        guard let takerInterfaceID = response[5] as? Data else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "takerInterfaceID"))
        }
        guard let stablecoin = response[6] as? EthereumAddress else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "stablecoin"))
        }
        guard let amountLowerBound = response[7] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "amountLowerBound"))
        }
        guard let amountUpperBound = response[8] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "taker"))
        }
        guard let securityDepositAmount = response[9] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "securityDepositAmount"))
        }
        guard let takenSwapAmount = response[10] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "takenSwapAmount"))
        }
        guard let serviceFeeAmount = response[11] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "serviceFeeAmount"))
        }
        guard let serviceFeeRate = response[12] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "serviceFeeRate"))
        }
        guard let direction = response[13] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "direction"))
        }
        guard let settlementMethod = response[14] as? Data else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "settlementMethod"))
        }
        guard let protocolVersion = response[15] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "protocolVersion"))
        }
        guard let isPaymentSent = response[16] as? Bool else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "isPaymentSent"))
        }
        guard let isPaymentReceived = response[17] as? Bool else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "isPaymentReceived"))
        }
        guard let hasBuyerClosed = response[18] as? Bool else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "hasBuyerClosed"))
        }
        guard let hasSellerClosed = response[19] as? Bool else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "hasSellerClosed"))
        }
        guard let disputeRaiser = response[20] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "disputeRaiser"))
        }
        return SwapStruct(isCreated: isCreated, requiresFill: requiresFill, maker: maker, makerInterfaceID: makerInterfaceID, taker: taker, takerInterfaceID: takerInterfaceID, stablecoin: stablecoin, amountLowerBound: amountLowerBound, amountUpperBound: amountUpperBound, securityDepositAmount: securityDepositAmount, takenSwapAmount: takenSwapAmount, serviceFeeAmount: serviceFeeAmount, serviceFeeRate: serviceFeeRate, direction: direction, settlementMethod: settlementMethod, protocolVersion: protocolVersion, isPaymentSent: isPaymentSent, isPaymentReceived: isPaymentReceived, hasBuyerClosed: hasBuyerClosed, hasSellerClosed: hasSellerClosed, disputeRaiser: disputeRaiser)
    }
    
    /**
     Called by `createFromGetSwapResponse` to get the appropriate message for a `BlockchainService.unexpectedNilError` when `createFromGetSwapResponse` is unable to get the value of a field that is expected in its `response` parameter.
     */
    private static func makeCreationErrorMessage(valueName: String) -> String {
        return "Got nil while getting \(valueName) value from getSwap response"
    }
    
    /**
     Returns an `Array` containing the properties of this struct, which can be passed to a web3swift write transaction call.
     */
    func toSwapDataArray() -> [AnyObject] {
        return [
            isCreated,
            requiresFill,
            maker,
            makerInterfaceID,
            taker,
            takerInterfaceID,
            stablecoin,
            amountLowerBound,
            amountUpperBound,
            securityDepositAmount,
            takenSwapAmount,
            serviceFeeAmount,
            serviceFeeRate,
            direction,
            settlementMethod,
            protocolVersion,
            isPaymentSent,
            isPaymentReceived,
            hasBuyerClosed,
            hasSellerClosed,
            disputeRaiser,
        ] as [AnyObject]
    }
}

File: androidApp/src/main/java/com/commuto/interfacemobile/android/blockchain/structs/SwapStruct.kt
package com.commuto.interfacemobile.android.blockchain.structs

import com.commuto.interfacemobile.android.contractwrapper.CommutoSwap
import java.math.BigInteger

/**
 * Represents an on-chain [Swap](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#swap) struct.
 *
 * @param isCreated Corresponds to an on-chain Swap's `isCreated` property.
 * @param requiresFill Corresponds to an on-chain Swap's `requiresFill` property.
 * @param maker Corresponds to an on-chain Swap's `maker` property.
 * @param makerInterfaceID Corresponds to an on-chain Swap's `makerInterfaceID` property.
 * @param taker Corresponds to an on-chain Swap's `taker` property.
 * @param takerInterfaceID Corresponds to an on-chain Swap's `takerInterfaceID` property.
 * @param stablecoin Corresponds to an on-chain Swap's `stablecoin` property.
 * @param amountLowerBound Corresponds to an on-chain Swap's `amountLowerBound` property.
 * @param amountUpperBound Corresponds to an on-chain Swap's `amountUpperBound` property.
 * @param securityDepositAmount Corresponds to an on-chain Swap's `securityDepositAmount` property.
 * @param takenSwapAmount Corresponds to an on-chain Swap's `takenSwapAmount` property.
 * @param serviceFeeAmount Corresponds to an on-chain Swap's `serviceFeeAmount` property.
 * @param serviceFeeRate Corresponds to an on-chain Swap's `serviceFeeRate` property.
 * @param direction Corresponds to an on-chain Swap's `direction` property.
 * @param settlementMethod Corresponds to an on-chain Swap's `settlementMethod` property.
 * @param protocolVersion Corresponds to an on-chain Swap's `protocolVersion` property.
 * @param isPaymentSent Corresponds to an on-chain Swap's `isPaymentSent` property.
 * @param isPaymentReceived Corresponds to an on-chain Swap's `isPaymentReceived` property.
 * @param hasBuyerClosed Corresponds to an on-chain Swap's `hasBuyerClosed` property.
 * @param hasSellerClosed Corresponds to an on-chain Swap's `hasSellerClosed` property.
 * @param disputeRaiser Corresponds to an on-chain Swap's `disputeRaiser` property.
 * @param chainID Corresponds to an on-chain Swap's `chainID` property.
 */
data class SwapStruct(
    val isCreated: Boolean,
    val requiresFill: Boolean,
    val maker: String,
    val makerInterfaceID: ByteArray,
    val taker: String,
    val takerInterfaceID: ByteArray,
    val stablecoin: String,
    val amountLowerBound: BigInteger,
    val amountUpperBound: BigInteger,
    val securityDepositAmount: BigInteger,
    val takenSwapAmount: BigInteger,
    val serviceFeeAmount: BigInteger,
    val serviceFeeRate: BigInteger,
    val direction: BigInteger,
    val settlementMethod: ByteArray,
    val protocolVersion: BigInteger,
    val isPaymentSent: Boolean,
    val isPaymentReceived: Boolean,
    val hasBuyerClosed: Boolean,
    val hasSellerClosed: Boolean,
    val disputeRaiser: BigInteger,
    val chainID: BigInteger,
) {

    /**
     * Returns a [CommutoSwap.Swap] derived from this [SwapStruct].
     *
     * @return A [CommutoSwap.Swap] derived from this [SwapStruct]
     */
    fun toCommutoSwapSwap(): CommutoSwap.Swap {
        return CommutoSwap.Swap(
            this.isCreated,
            this.requiresFill,
            this.maker,
            this.makerInterfaceID,
            this.taker,
            this.takerInterfaceID,
            this.stablecoin,
            this.amountLowerBound,
            this.amountUpperBound,
            this.securityDepositAmount,
            this.takenSwapAmount,
            this.serviceFeeAmount,
            this.serviceFeeRate,
            this.direction,
            this.settlementMethod,
            this.protocolVersion,
            this.isPaymentSent,
            this.isPaymentReceived,
            this.hasBuyerClosed,
            this.hasSellerClosed,
            this.disputeRaiser,
        )
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as SwapStruct

        if (isCreated != other.isCreated) return false
        if (requiresFill != other.requiresFill) return false
        if (maker != other.maker) return false
        if (!makerInterfaceID.contentEquals(other.makerInterfaceID)) return false
        if (taker != other.taker) return false
        if (!takerInterfaceID.contentEquals(other.takerInterfaceID)) return false
        if (stablecoin != other.stablecoin) return false
        if (amountLowerBound != other.amountLowerBound) return false
        if (amountUpperBound != other.amountUpperBound) return false
        if (securityDepositAmount != other.securityDepositAmount) return false
        if (takenSwapAmount != other.takenSwapAmount) return false
        if (serviceFeeAmount != other.serviceFeeAmount) return false
        if (serviceFeeRate != other.serviceFeeRate) return false
        if (direction != other.direction) return false
        if (!settlementMethod.contentEquals(other.settlementMethod)) return false
        if (protocolVersion != other.protocolVersion) return false
        if (isPaymentSent != other.isPaymentSent) return false
        if (isPaymentReceived != other.isPaymentReceived) return false
        if (hasBuyerClosed != other.hasBuyerClosed) return false
        if (hasSellerClosed != other.hasSellerClosed) return false
        if (disputeRaiser != other.disputeRaiser) return false
        if (chainID != other.chainID) return false

        return true
    }

    override fun hashCode(): Int {
        var result = isCreated.hashCode()
        result = 31 * result + requiresFill.hashCode()
        result = 31 * result + maker.hashCode()
        result = 31 * result + makerInterfaceID.contentHashCode()
        result = 31 * result + taker.hashCode()
        result = 31 * result + takerInterfaceID.contentHashCode()
        result = 31 * result + stablecoin.hashCode()
        result = 31 * result + amountLowerBound.hashCode()
        result = 31 * result + amountUpperBound.hashCode()
        result = 31 * result + securityDepositAmount.hashCode()
        result = 31 * result + takenSwapAmount.hashCode()
        result = 31 * result + serviceFeeAmount.hashCode()
        result = 31 * result + serviceFeeRate.hashCode()
        result = 31 * result + direction.hashCode()
        result = 31 * result + settlementMethod.contentHashCode()
        result = 31 * result + protocolVersion.hashCode()
        result = 31 * result + isPaymentSent.hashCode()
        result = 31 * result + isPaymentReceived.hashCode()
        result = 31 * result + hasBuyerClosed.hashCode()
        result = 31 * result + hasSellerClosed.hashCode()
        result = 31 * result + disputeRaiser.hashCode()
        result = 31 * result + chainID.hashCode()
        return result
    }
}